Moseq Pipeline

samuelbray32 opened the pull request:
# Description
### MoSeq Pipeline Support:
- Implementation of moseq within spyglass pipelines.
    -  Keypoint data sourced from `PositionOutput`
        - Added `PositionOutput.get_pose_dataframe` to allow fetching of full pose rather than centroid for appropriate pipelines.
    - `PoseGroup` used to combine multiple epochs of pose data into a training set (analagous to `PositionGroup` in decoding pipeline) 
    - Model parameters and training arguments stored in `MoseqModelParams`
        - Can define initial model as an existing entry in `MoseqModel` for reuse and extended training
    - `MoseqModelSelection` combines `PoseGroup` andmodel parameters
    - `MoseqModel` trains model and stores in external project file
    - `MoseqSyllable` (and it's selction table) apply a trained model to an entry from `PositionOutput`
        - stores syllables and latent variables as a dataframe in an `AnalysisNwbfile` entry

- Added utility function `get_position_interval_epoch`

### MoSeq Pipeline Support:
* Added new environment files `environment_moseq_cpu.yml` and `environment_moseq_gpu.yml` for setting up MoSeq environments. [[1]](diffhunk://#diff-71726924bad0860580295bbe74d03034c39b0b2c986e4052f6fd40fbbfee8b85R1-R42) [[2]](diffhunk://#diff-9f61b71c15a5f701f051d2f7ebb76942e01a64acd688d0927ccb0b9a8b93f0c7R1-R42)
* Updated `pyproject.toml` to include MoSeq dependencies for both CPU and GPU setups. [[1]](diffhunk://#diff-50c86b7ed8ac2cf95bd48334961bf0530cdc77b5a56f852c5c61b89d735fd711R68-R78) [[2]](diffhunk://#diff-50c86b7ed8ac2cf95bd48334961bf0530cdc77b5a56f852c5c61b89d735fd711R191)

### Documentation Updates:
* Enhanced `notebooks/00_Setup.ipynb` and `notebooks/py_scripts/00_Setup.py` with instructions for setting up Deep Lab Cut (DLC) and MoSeq environments. [[1]](diffhunk://#diff-374712d89b7c1e8b5604afd6dd12452f09a00e44fc89fdc8f7139e73692fe4a5L288-R328) [[2]](diffhunk://#diff-374712d89b7c1e8b5604afd6dd12452f09a00e44fc89fdc8f7139e73692fe4a5L596-R636) [[3]](diffhunk://#diff-374712d89b7c1e8b5604afd6dd12452f09a00e44fc89fdc8f7139e73692fe4a5L610-R650) [[4]](diffhunk://#diff-4a7c9174e232b23076d732a9a77c683d4c55ea609c03db8e0df77af5af1ad381L10-R10) [[5]](diffhunk://#diff-4a7c9174e232b23076d732a9a77c683d4c55ea609c03db8e0df77af5af1ad381R260-R299)
* Added a new tutorial notebook `notebooks/py_scripts/60_MoSeq.py` for the MoSeq pipeline, providing detailed steps for model training and analysis.

### Configuration Updates:
* Updated `dj_local_conf_example.json` to include new directories for MoSeq data.

### Changelog:
* Updated `CHANGELOG.md` to reflect the addition of the common method `get_position_interval_epoch` and the implementation of the keypoint-moseq extraction pipeline. [[1]](diffhunk://#diff-06572a96a58dc510037d5efa622f9bec8519bc1beab13c9f251e97e657a9d4edR9) [[2]](diffhunk://#diff-06572a96a58dc510037d5efa622f9bec8519bc1beab13c9f251e97e657a9d4edR26-R29)

# My To-dos
- [x] improve the tutorial with more explanations of the steps
- [x] update notebook image to use `PositionOutput`
- [x] change the model stored directories to the spyglass config
- [x] make a better standard version for getting epoch from the position interval name (as mentioned by Chris before)
- [x] update changelog

# Checklist:

<!--
For items below with choices, select one (e.g., yes, no) 
and check the box to indicate that a choice was selected.
If not applicable, check the box and add `N/A` after the box.
For example:
- [X] This has a choice: no
- [X] N/A. If choice, other item.
This comment may be deleted on submission.

Alter notes example:
```python
from spyglass.example import Table
Table.alter() # Comment regarding the change
```
-->

- [x] N This PR should be accompanied by a release: (yes/no/unsure)
- [x] NA If release, I have updated the `CITATION.cff`
- [x] N This PR makes edits to table definitions: (yes/no)
- [x] NA If table edits, I have included an `alter` snippet for release notes.
- [x] NA If this PR makes changes to position, I ran the relevant tests locally.
- [x] I have updated the `CHANGELOG.md` with PR number and description.
- [x] I have added/edited docs/notebooks to reflect the changes

CBroz1 commented on src/spyglass/behavior/moseq.py:
```diff
@@ -0,0 +1,360 @@
+        # increment param name
+        if new_name is None:
+            # increment the extension number
+            if model_key["model_params_name"][:-1].endswith("_extension"):
```
if this moseq's convention? using the `extension` suffix? I'd avoid mixing data types like this if possible

samuelbray32 replied:
This is part of a system I made to allow for extended training of an existing model (defined by `model key`)

There isn't anything particular to the term extension, but it allows me to have this helper function for users to build an extended traning

samuelbray32 replied:
That's not a bad idea, by I don't know that we need to track it that strictly:

My logic was to be as minimally prescriptive on structure here since rounds of training can be a variable process. The helper function here is to extend training of a model with the same parameters, but you could also extend training having changed the parameters.  An in both these cases you could be using the same or different training data from `PoseGroup`. In some of these cases when extension should reset to zero is mostly just a convention choice.

Also, the naming structure here doesn't do anything functionally for tracking provenance (that's done by the `initial_model` key in the params) and users can pass different names to it that are useful to them.  

To minimize things the things the user has to pay attention to, I think it makes sense to just use the `model_params_name` column.  I'll try to think of a different naming scheme that feels less awkward here though

CBroz1 commented on src/spyglass/behavior/moseq.py:
```diff
@@ -0,0 +1,360 @@
+
+
+@schema
+class MoseqModelParams(SpyglassMixin, dj.Manual):
```
I don't love the convention of table names prefixed by the software dependency (`DLCModelX`, `MoseqModelX`). It permits

```python
from spyglass.behavior.moseq import *
from spyglass.position.whatever import *

DLCModel()
MoseqModel()
```

I'd prefer to encourage

```python
from spyglass.behavior import moseq
from spyglass.position.whatever import dlc_model as dlc

moseq.Model()
dlc.Model()
```

but it's hard to shift to the latter when we already have the former

CBroz1 commented on src/spyglass/behavior/moseq.py:
```diff
@@ -0,0 +1,360 @@
+    """
+
+    def make_training_extension_params(
+        self, model_key, num_epochs, new_name=None, skip_duplicates=False
```
```suggestion
        self, model_key:dict, num_epochs:int, new_name:str=None, skip_duplicates:bool=False
```

CBroz1 commented on src/spyglass/behavior/moseq.py:
```diff
@@ -0,0 +1,360 @@
+        # either initialize a new model or load an existing one
+        model_name = self._make_model_name(key)
+        initial_model_key = model_params.get("initial_model", None)
+        if initial_model_key is None:
```
Initializing a model seems like a good candidate for a private method

CBroz1 commented on src/spyglass/behavior/moseq.py:
```diff
@@ -0,0 +1,360 @@
+            key = {}
+        key = (MoseqModelSelection & key).fetch1("KEY")
+        key_string = "_".join(key.values())
+        return hashlib.sha1(key_string.encode("utf-8")).hexdigest()[:10]
```
I'd prefer to use datajoint's builtin hasher in these cases, rather than make a new one to handle unhashable types like dict

`dj.hash.key_hash`

CBroz1 commented on src/spyglass/behavior/moseq.py:
```diff
@@ -0,0 +1,360 @@
+        ----------
+        key : dict
+            key to a single MoseqModel table entry
+
+        Raises
+        ------
+        NotImplementedError
+            This method is not implemented yet.
```
```suggestion
```

CBroz1 commented on src/spyglass/behavior/moseq.py:
```diff
@@ -0,0 +1,360 @@
+        """
+        if key is None:
+            key = {}
+        model = kpms.load_checkpoint(
+            (self & key).fetch1("project_dir"),
+            (self & key).fetch1("model_name"),
+        )[0]
+        return model
```
```suggestion
        return kpms.load_checkpoint(
            (self & key).fetch1("project_dir"),
            (self & key).fetch1("model_name"),
        )[0]
```

CBroz1 reviewed (commented):
Took a first pass on the behavior subpackage - Can continue review tomorrow

edeno commented on src/spyglass/behavior/core.py:
```diff
@@ -0,0 +1,200 @@
+    def create_group(
+        self,
+        group_name: str,
+        merge_ids: list[str],
```
I think actually depends on the version of python. But probably true for the version of python we are using.

CBroz1 commented on src/spyglass/position/v1/position_dlc_selection.py:
```diff
@@ -65,69 +65,84 @@ def make(self, key):
-            comments=pos_obj.comments,
-            description=pos_obj.description,
-        )
+        if isinstance(pos_nwb["dlc_position"], pd.DataFrame):
```
I think I would migrate these to a separate method to clean up the indenting, something like `declare_pos_analysis_file`

samuelbray32 replied:
@CBroz1, reorganized this make function. Let me know if it looks cleaner to you

CBroz1 commented on src/spyglass/position/v1/position_dlc_pose_estimation.py:
```diff
@@ -406,6 +406,16 @@ def fetch_dataframe(self, *attrs, **kwargs):
             axis=1,
         )
 
+    def fetch_video_name(self):
```
This name doesn't totally align with what's going - it fetches the path instead of the name. I think the calling func then applies `stem` to extract name from path. I think you could either change this to `fetch_video_path` or add `stem` to the func body

samuelbray32 replied:
changed the name. thanks

CBroz1 commented on src/spyglass/position/v1/position_trodes_position.py:
```diff
@@ -267,7 +278,7 @@ def make(self, key):
             )
             + 1
-        )
+        )  # TODO: Fix this hack
```
There are quite a few places in the code where we substring `pos X valid times` - def worth revising/unifying the approach
- `position_dlc_selection.pyL333`
- `common_position.pyL544`
- `common_interval.pyL133` uses regex, which is a little better
- `common_behav.py.PositionSource` has some helper funcs for compose/decompose

samuelbray32 replied:
agreed.  The correct solution now would use a join of `PositionIntervalMap * TaskEpoch` which is what I do in my personal code.  Would make sense to put this in a common function

CBroz1 reviewed (commented):
Looking good so far!

samuelbray32 commented:
I've now merged in the main branch and have tested it working with a dependency on `PositionOutput` rather than the introduced `PoseOutput` table proposed before. Remaining to-do's:

- [ ] improve the tutorial with more explanations of the steps
- [ ] change the model stored directories to the spyglass config
- [ ] make a better standard version for getting epoch from the position interval name (as mentioned by Chris before)

edeno reviewed (changes requested):
Can you add the dependencies to the pyproject.toml and environment.yml?

samuelbray32 commented:
> Can you add the dependencies to the pyproject.toml and environment.yml?

Yeah.  Do we want them in the default dependencies or should they be their own version similar to how we do the DLC ones?

CBroz1 commented on src/spyglass/position/v1/position_dlc_pose_estimation.py:
```diff
@@ -409,6 +409,16 @@ def fetch_dataframe(self, *attrs, **kwargs) -> pd.DataFrame:
         )
 
+    def fetch_video_path(self):
+        """Return the video path for pose estimate
```
Normally, we pass key as an arg for this kind of thing. It's not clear to me why you break the pattern here. If no arg, it could be a property, yeah? If kept, I'd add example use to the docstring

```suggestion
        """Return the video path for pose estimate
        
        >>> (DLCPoseEstimation & key).fetch_video_path()
```

CBroz1 commented on src/spyglass/position/v1/position_dlc_selection.py:
```diff
@@ -269,6 +294,27 @@ def evaluate_pose_estimation(cls, key):
+
+    def fetch_video_path(self):
+        """Return the video path for pose estimate
+
```
See above note on break from key-as-arg convention

CBroz1 commented on src/spyglass/position/v1/position_dlc_selection.py:
```diff
@@ -269,6 +294,27 @@ def evaluate_pose_estimation(cls, key):
 
+    def fetch_pose_dataframe(self):
+        """fetches the pose data from the pose estimation table
+
```
See above note on break from key-as-arg convention

CBroz1 reviewed (commented):
I did another pass here with some comments. Are pytests in the scope of this PR, or do we want to hold off?

CBroz1 commented:
> Yeah. Do we want them in the default dependencies or should they be their own version similar to how we do the DLC ones?

A user will need a DLC project in place to use MoSeq, right? I think I might rename our 'dlc' optional dependency to 'position' and moseq dependencies in there - not sure which I prefer across the following:

```
pip install .[dlc,moseq] # separate categories in pyproject
pip install .[position] 
pip install .[pos_deps]
```

samuelbray32 commented on src/spyglass/behavior/moseq.py:
```diff
@@ -0,0 +1,374 @@
+    model_params_name: varchar(80)
+    ---
+    model_params: longblob
+    """
```
I think variability of the skeleton make calling any specific one a "default" is hard.  Do you think including the tutorial one here (and naming it such) would make sense or no?

samuelbray32 commented on src/spyglass/behavior/moseq.py:
```diff
@@ -0,0 +1,374 @@
+        video_paths = (PoseGroup & key).fetch_video_paths()
+        for video in video_paths:
+            destination = os.path.join(video_dir, os.path.basename(video))
+            if not os.path.lexists(destination):
```
Currently, nothing in the pipeline uses the video, so that case wouldn't raise errors. Agree it's non-ideal, I'll make this more robust

samuelbray32 commented on src/spyglass/behavior/moseq.py:
```diff
@@ -0,0 +1,374 @@
+            }
+        )
+
+    def _config_func(self, project_dir):
+        return lambda: kpms.load_config(project_dir)
```
For some reason I had it in my head that the `kpms` module needed the config as a callable function in places, but that was only in methods I implemented. I'll clean that up

samuelbray32 commented on src/spyglass/common/common_behav.py:
```diff
@@ -686,6 +686,30 @@ def get_interval_list_name_from_epoch(nwb_file_name: str, epoch: int) -> str:
+    ):
+        convert_epoch_interval_name_to_position_interval_name(epoch_key)
+    # try again
+    query = PositionIntervalMap * TaskEpoch & key
+    if query:
+        return query.fetch1("epoch")
+    return None
```
Shouldn't need to try more than twice.  The second attempt is just after ensuring that all epoch intervals have been mapped to a position interval.  That wouldn't change with further attempts

samuelbray32 commented:
> > Yeah. Do we want them in the default dependencies or should they be their own version similar to how we do the DLC ones?
> 
> A user will need a DLC project in place to use MoSeq, right? I think I might rename our 'dlc' optional dependency to 'position' and moseq dependencies in there - not sure which I prefer across the following:
> 
> ```
> pip install .[dlc,moseq] # separate categories in pyproject
> pip install .[position] 
> pip install .[pos_deps]
> ```

I may lean towards keeping them separate. You don't strictly need DLC in your environment to run the moseq pipeline.
Examples:
- User 2 is pulling pose data from a pre-existing DLC entry made by a different User 1
- I think it would be reasonable future enhancement to have some version of a `ImportedPose` table to allow new users to bring pre-existing DLC annotated data into spyglass (analogous to `ImportedSpikeSorting`)

edeno commented on pyproject.toml:
```diff
@@ -65,6 +65,17 @@ dlc = [
+    "jax-moseq",
+    "keypoint-moseq",
+]
+
```
Update the `omit` section because this module is currently untested.

edeno reviewed (changes requested):
I ran through the notebook and ran into a couple of problems. I was unable to run the moseq algorithm I think but it was unclear to me why.

samuelbray32 commented:
<p>When I deleted and ran this again I had an issue related due to the project path.  I changed the code to make it unique prior to creation of the moseq model</p><p> </p>

---

 <a href='https://app.reviewnb.com/LorenFrankLab/spyglass/pull/1056/discussion/#comment-2603189928'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

edeno commented:
Ran into an issue in the notebook
<img width="766" alt="image" src="https://github.com/user-attachments/assets/fd3864ec-cd9d-4477-a400-2ad8ea49e643" />

samuelbray32 commented:
@edeno 

Not sure why that would still raise an error. I added a default value for the passed key [here](https://github.com/LorenFrankLab/spyglass/blob/d68890aa8b59cf713c1df89092b054d97dae96a8/src/spyglass/behavior/v1/moseq.py#L254) in commit [efeaf0d](https://github.com/LorenFrankLab/spyglass/pull/1056/commits/efeaf0dd48b56bf4e23a8b9759f948decb5c9600). Is it possible that didn't get pulled?

edeno commented:
I definitely pulled it. I'll let @CBroz1 go through it and see if he encounters any similar problems? I suppose it could be related to my installation of moseq?

CBroz1 commented on src/spyglass/position/position_merge.py:
```diff
@@ -75,3 +77,23 @@ def fetch1_dataframe(self) -> DataFrame:
         return query.fetch1_dataframe()
+
+    def fetch_pose_dataframe(self):
+        key = self.merge_restrict(self.proj()).proj()
```
I think this returns a table, rather than a restriction. In most cases, that'll work, but it causes issues in the export process, because export tries to log a table as a restriction. The following is the easiest way around it, but maybe there's a better solution
```suggestion
        key = self.merge_restrict(self.proj()).proj().restriction
```

samuelbray32 replied:
Agree with the change here.  In addition, could we check if `is_instance(restriction,dj.FreeTable)` in the export logging processand cast it to a restriction in the export process? Just to catch cases like this in user code in the future

CBroz1 replied:
Good idea

CBroz1 replied:
Correction here: `.restriction` won't work. It has to be `.fetch('KEY', as_dict=True)` to avoid the same subquery issue

CBroz1 replied:
What I landed on in my editing:

```python
key = self.merge_restrict(self.fetch("KEY", as_dict=True)).fetch(
    "KEY", as_dict=True)
)
```

CBroz1 reviewed (changes requested):
I spent some time trying to run an export so I could start from PositionOutput and run moseq locally (with various resulting bug reports). I wanted to submit the handful of comments I had before the weekend. I'm hoping I'll be able to get the export working on monday to actually be able to run the moseq pipeline

CBroz1 commented on src/spyglass/position/v1/position_dlc_pose_estimation.py:
```diff
@@ -408,6 +408,21 @@ def fetch_dataframe(self, *attrs, **kwargs) -> pd.DataFrame:
+        str
+            absolute path to video file
+        """
+        key = (self & key).fetch1("KEY")
```
```suggestion
        key = (self & key).fetch1("KEY", log_export=False)
```

CBroz1 reviewed (changes requested):
I'm able to run the notebook locally, so I'm good to merge whenever there's time to address my outstanding comments. Testing would be the most major one, for sure, but we can kick the can

edeno commented on src/spyglass/settings.py:
```diff
@@ -185,12 +201,14 @@ def load_config(
-                    if prefix == "dlc"
-                    else dj_kachery if prefix == "kachery" else dj_spyglass
-                )
+                if prefix == "dlc":
```
This is probably better as a dictionary with this many if statements

samuelbray32 commented on src/spyglass/behavior/v1/core.py:
```diff
@@ -0,0 +1,199 @@
+from spyglass.position.position_merge import PositionOutput
+from spyglass.utils import SpyglassMixin, SpyglassMixinPart
+
+schema = dj.schema("behavior_v1_core")
```
added `behavior` to the shared schemas list
changed the moseq schema name to `behavior_v1_moseq`

samuelbray32 commented:
I think I've cleared all the suggestions and comments. 

For tests, I would prefer to merge this PR in and then work on those as a new to-do item. Just to reduce mental load.

CBroz1 commented on src/spyglass/position/position_merge.py:
```diff
@@ -79,7 +79,13 @@ def fetch1_dataframe(self) -> DataFrame:
 
     def fetch_pose_dataframe(self):
-        key = self.merge_restrict(self.proj()).proj()
+        """
+        Fetch a single dataframe of pose bodypart coordinates.
+        This will work only for sources that have pose information.
+        """
```
The [dostring style guide](https://numpydoc.readthedocs.io/en/latest/format.html) spyglass follows recommends separating the oneliner from the rest.
```suggestion
        """Fetch a single dataframe of pose bodypart coordinates.
        
        This will work only for sources that have pose information.
        """
```

CBroz1 reviewed (approved):
One minor comment, but ready to merge
